EFCache.Redis/RedisCache.cs

```diff
@@ -167,75 +170,105 @@ public bool GetItem(string key, out object value)
+				transaction.SetRemoveAsync(entitySetKey, hashedKey);
 			}
-			catch (Exception e)
+
+			var committed = false;
+			PerformWithRetryBackoff(() =>
 			{
-				OnCachingFailed(e);
+				committed = transaction.Execute();
+			});
+
+			if (!ShouldCollectStatistics || !committed) return;
+
+			foreach (var (key, _) in keysAndEntitySetsToInvalidate)
+			{
+				database.HashIncrement(AddStatsQualifier(HashKey(key)), InvalidationsIdentifier, 1, CommandFlags.FireAndForget);
 			}
 		}
 
+		// Note: This method does not use try/catch. It attempts retry and then throws if unsuccessful
+		// This method should never be allowed to proceed with an exception as it will lead to
+		// an incoherent cache
 		public void InvalidateItem(string key)
 		{
 			key.GuardAgainstNullOrEmpty(nameof(key));
 
 			var database = Connection.GetDatabase();
 			var hashedKey = HashKey(key);
 
-			try
+			var entry = database.ObjectGet<CacheEntry>(hashedKey);
+			if (entry == null) return;
+
+			var transaction = database.CreateTransaction();
+			transaction.KeyDeleteAsync(hashedKey);
+			foreach (var set in entry.EntitySets)
 			{
-				var entry = database.ObjectGet<CacheEntry>(hashedKey);
-				if (entry == null) return;
+				transaction.SetRemoveAsync(AddCacheQualifier(set), hashedKey);
+			}
 
-				var transaction = database.CreateTransaction();
-				transaction.KeyDeleteAsync(hashedKey);
-				foreach (var set in entry.EntitySets)
-				{
-					transaction.SetRemoveAsync(AddCacheQualifier(set), hashedKey);
-				}
-				var committed = transaction.Execute();
+			var committed = false;
+			PerformWithRetryBackoff(() =>
+			{
+				committed = transaction.Execute();
+			});
 
-				if (!ShouldCollectStatistics || !committed) return;
-				
-				database.HashIncrement(AddStatsQualifier(hashedKey), InvalidationsIdentifier, 1, CommandFlags.FireAndForget);
-			}
-			catch (Exception e)
+			if (!ShouldCollectStatistics || !committed) return;
+
+			database.HashIncrement(AddStatsQualifier(hashedKey), InvalidationsIdentifier, 1, CommandFlags.FireAndForget);
+		}
+
+		/// <summary>
+		/// This provides a retry mechanism and throws without handling if retries are not successful.
+		/// This should only be used by the invalidation process. Gets and sets should be quick and
+		/// there are no data integrity issues if they fail.
+		/// An unsuccessful invalidation will cause the cache to become incoherent and should be avoided
+		/// at all costs.
+		/// </summary>
+		/// <param name="action"></param>
+		private static void PerformWithRetryBackoff(Action action)
+		{
+			Exception aggregateException = new RedisCacheException("Retry exception base exception");
+			// Attempt the action, plus up to RetryLimit additional attempts
+			for (var i = 0; i < RetryLimit; i++)
 			{
-				OnCachingFailed(e);
+				try
+				{
+					action();
+					return;
+				}
+				catch (Exception e)
+				{
+					aggregateException = new Exception("Retry exception", new Exception("Action exception", e));
```

Comment:
@jkdufair were you meaning to combine all the exceptions together, or only track the last failure? i think this just overrides the exception every loop iteration.

Reply:
The former. Tried it until my eyes crossed. I'll take another look LOL.